src/meta/src/hummock/compaction/picker/compaction_task_validator.rs

```diff
@@ -0,0 +1,217 @@
+                    .table_infos
+                    .iter()
+                    .map(|sst| sst.file_size)
+                    .sum::<u64>()
+            })
+            .sum();
+
+        let compact_file_count: u64 = input
+            .input_levels
+            .iter()
+            .map(|overlapping_sub_level| overlapping_sub_level.table_infos.len() as u64)
+            .sum();
+
+        // Limit sstable file count to avoid using too much memory.
+        let overlapping_max_compact_file_numer = std::cmp::min(
+            self.config.level0_max_compact_file_number,
+            MAX_COMPACT_LEVEL_COUNT as u64,
+        );
+
+        let waiting_enough_files = {
+            if compaction_bytes > max_compaction_bytes {
+                false
+            } else {
+                compact_file_count <= overlapping_max_compact_file_numer
+            }
+        };
+
+        // If waiting_enough_files is not satisfied, we will raise the priority of the number of
+        // levels to ensure that we can merge as many sub_levels as possible
+        let tier_sub_level_compact_level_count =
+            self.config.level0_overlapping_sub_level_compact_level_count as usize;
+        if input.input_levels.len() < tier_sub_level_compact_level_count && waiting_enough_files {
+            stats.skip_by_count_limit += 1;
+            return false;
+        }
+
+        true
+    }
+
+    fn valid_intra_compaction(
+        &self,
+        input: &CompactionInput,
+        stats: &mut LocalPickerStatistic,
+    ) -> bool {
+        let intra_sub_level_compact_level_count =
+            self.config.level0_sub_level_compact_level_count as usize;
+
+        if input.input_levels.len() < intra_sub_level_compact_level_count {
+            return false;
+        }
+
+        let mut max_level_size = 0;
+        let mut total_file_size = 0;
+        let mut total_file_count = 0;
+        for select_level in &input.input_levels {
+            let level_select_size = select_level
+                .table_infos
+                .iter()
+                .map(|sst| sst.file_size)
+                .sum::<u64>();
+
+            max_level_size = std::cmp::max(max_level_size, level_select_size);
+
+            total_file_size += level_select_size;
+            total_file_count += select_level.table_infos.len();
+        }
+
+        // This limitation would keep our write-amplification no more than
+        // ln(max_compaction_bytes/flush_level_bytes) /
+        // ln(self.config.level0_sub_level_compact_level_count/2) Here we only use half
+        // of level0_sub_level_compact_level_count just for convenient.
+        let is_write_amp_large =
+            max_level_size * self.config.level0_sub_level_compact_level_count as u64 / 2
+                >= total_file_size;
+
+        if is_write_amp_large
+            && total_file_count < self.config.level0_max_compact_file_number as usize
+        {
+            stats.skip_by_write_amp_limit += 1;
```

Comment:
I think updating these skip_by_* in `valid_compact_task` is improper, because `valid_compact_task` can be called multiple times for one pick.